flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -528,7 +535,11 @@ public void failSlot(
 
 	private void internalFailAllocation(AllocationID allocationId, Exception cause) {
 		final Optional<ResourceID> resourceIdOptional = slotPool.failAllocation(allocationId, cause);
-		resourceIdOptional.ifPresent(this::releaseEmptyTaskManager);
+		resourceIdOptional.ifPresent(taskManagerId -> {
+			if (!partitionTable.hasTrackedPartitions(taskManagerId)) {
```

Comment:
nit: if we put the if (!partitionTable.hasTrackedPartitions(taskManagerId)) inside the method of releaseEmptyTaskManager, we could reuse it if release might be reused in other parts future.